### ISSUE TITLE
Why is the model setup as it is, especially the final decoding process?

### ISSUE THREAD

**Issue by LexiconLookout** [facebookresearch/svoice#12]
Hi everyone, 

I've been trying to get a grip on machine learning and speech separation. So I thought it'd be a good exercise to manage to understand how this particular setup works. Initially, I did so by going over the paper about this model, (https://arxiv.org/pdf/2003.01531.pdf) but its descriptions proved somewhat ambiguous from time-to-time. Once I found the code on github, analyzing that answered most of my questions on the exact setup that was used. So I feel I have mostly a grip on the 'how'. What I can't seem to wrap my head around is the 'why', and with the answer eluding me when looking it up for myself, I thought I would just ask.

As the title mentioned, it is mostly the final decoding that mystifies me. Here's a longwinded description of how it goes, as far as I can follow along.
 
 ----------
 
-In the separator subclass, the forward function first calls the DPMulCat rnn which outputs the expanded blocks after they've been Prelu'd and through filter D. They are size N(filters) x CR (chunks for all voices) x K (time-series segments).

-These are then fed into MergeChunks function which transposes them (N x K x CR) and reshapes them. (N x 0.5K x 2CR) Through this reshaping process the first chunk of new_K x new_CR will have the top rows filled with the information from the first old_K x old_CR chunk, with the second one filling the rows at the bottom. So all the information is already somewhat scrambled.

-Then it takes the first half of columns in new_K x _new_CR, flattens this matrix and removes the first part, and sums it with the final half of columns in this matrix, also flattened but with the final part removed. It does so for all N filters, getting the desired form of information N x CT', but having it be even more scrambled.

-These tensors are passed on to the an SWave class object, which first reshapes them to have an extra dimension for the speakers, which then all have an N x T' matrix attributed to them. This would attribute the top rows in the N x CT' matrix to the first speaker, rows below that to the second and so on. This information is scrambled again.

-After this step, the information is passed to the decoder object which transposes it again (T' x N) then averages the values in the along the rows N by steps of the original kernel size L used to 1D-Convolve the audio signal. (T' x N/L) and only after this is the actual overlap_and_add function used along the column dimension of T' with step size of 1/2L to get back to a signal of approximately length T.

-This signal is then padded back in the SWave object to the original length, although I saw no means to deal with a signal that is longer than the original signal which could occur, I think, if N/L > L
 
 -------------
 
So I can follow along fairly well with what it does, mostly. But WHY it would be set up like this is a complete mystery to me. 

Wouldn't doing it like this have the learning model not only have to learn how to actually separate the signals, but also how to recode them to fit the decoding process, giving it double work and therefore making it learn slower? Or was this done on purpose to prevent the system from being 'lazy' during the learning process which somehow results in it being more accurate?

The same goes for the MULCAT blocks. Is there a justification for their structure? Is the structure of a mixed audio signal such that, {adding two LSTM blocks, putting a fully connected layer afterwards to get back your original size, then multiplying the results of the two element-wise, then concatenating it with the original input, then putting another fully connected layer afterwards to get to the original size, and then adding the original block element-wise, then transposing and doing this over-and-over,} was expected to improve the model over Luo, Y., Chen, Z., and Yoshioka, T. 2019c(arXiv:1910.06379)? Or was it just empirically determined by trying lots of setups?

**Comment by adiyoss** [comment]
Hi @LexiconLookout :),
Thanks for you interest in our research! 
If you have specific questions regarding the model arch. / setup / architecture I suggest to mail the authors rather than opening an issue on GitHub. 
In case you have questions about the code / something is not working, feel free to open another issue!

Thanks!
